### DISCUSSION TITLE
Does tB optimise out NOOPs?

### DISCUSSION THREAD

**Discussion by Greedquest** [twinbasic/twinbasic#1644] [Q&A]
I have lots of logging in my code. If I make the body of the log function a no-op is tB smart enough to eliminate that call?

A function call is fast but it's the parameters (copying strings about) that will be slow.

```vba
Sub Log(byval s as string)
    #if loggingEnabled Then
        Logmanager.Log s
    #end if
End sub
```

Will tB eliminate calls to this function? Or do I need conditional compilation at every call site?

**Comment by Greedquest** [comment]
I think LLVM will tackle this. If producing or passing the parameters is side affecting then even if the body is empty the function call is actually still "doing something" so cannot be eliminated IIUC.

**Reply by bclothier** [reply]
Even if it did, I still think a better case is made for using attributes rather than conditional compilation. See here:

https://github.com/twinbasic/twinbasic/issues/25#issuecomment-821741827
